File: packages/template-rust/src/render/runtime_environment.rs
//! Host functions and resource limits shared by compiler modes.

use super::context::{Limits, RuntimeServices};
use crate::functions::HostFunction;
use std::collections::HashMap;

/// Concrete runtime services owned by AST and generated programs.
pub struct RuntimeEnvironment {
    /// Active resource limits.
    pub limits: Limits,
    /// Registered host functions.
    pub host_functions: HashMap<String, HostFunction>,
    /// Registered logical class functions.
    pub class_functions: HashMap<String, HostFunction>,
}

impl RuntimeEnvironment {
    /// Creates runtime services from explicit limits and host functions.
    pub fn new(limits: Option<Limits>, host_functions: HashMap<String, HostFunction>) -> RuntimeEnvironment {
        RuntimeEnvironment {
            limits: limits.unwrap_or_default(),
            host_functions,
            class_functions: HashMap::new(),
        }
    }

    /// Registers one logical class function.
    pub fn register_class(&mut self, class_name: &str, method: &str, function: HostFunction) -> Result<(), String> {
        let valid = |name: &str| {
            name.bytes().next().is_some_and(|byte| byte.is_ascii_alphabetic() || byte == b'_')
                && name.bytes().all(|byte| byte.is_ascii_alphanumeric() || byte == b'_')
        };
        if !valid(class_name) || !valid(method) {
            return Err("class function names must be identifiers".to_string());
        }
        self.class_functions.insert(format!("{class_name}::{method}"), function);
        Ok(())
    }

    /// Registers a host function after validating its name.
    pub fn register(&mut self, name: &str, function: HostFunction) -> Result<(), String> {
        let valid = name.bytes().next().is_some_and(|byte| byte.is_ascii_alphabetic() || byte == b'_')
            && name.bytes().all(|byte| byte.is_ascii_alphanumeric() || byte == b'_');
        if !valid {
            return Err(format!("{name:?} is not an identifier"));
        }
        if crate::functions::builtins().contains_key(name) {
            return Err(format!("{name} is a built-in function"));
        }
        self.host_functions.insert(name.to_string(), function);
        Ok(())
    }
}

impl RuntimeServices for RuntimeEnvironment {
    fn limits(&self) -> Limits {
        self.limits
    }

    fn host_function(&self, name: &str) -> Option<&HostFunction> {
        self.host_functions.get(name)
    }

    fn class_function(&self, class_name: &str, method: &str) -> Option<&HostFunction> {
        self.class_functions.get(&format!("{class_name}::{method}"))
    }
}

File: packages/template-ts/src/render/runtime-environment.ts
// Runtime environment shared by AST and generated programs.
import { builtins, type HostFunction } from '../functions/index.js';
import { DEFAULT_LIMITS, type Limits, type RuntimeServices } from './context.js';

/** Host functions and resource limits shared by AST and generated programs. */
export class RuntimeEnvironment implements RuntimeServices {
  private readonly runtimeLimits: Limits;
  private readonly hostFunctions = new Map<string, HostFunction>();
  private readonly classFunctions = new Map<string, HostFunction>();

  /** Creates an environment from resource limits and validated host functions. */
  constructor(limits: Partial<Limits> = {}, functions: Record<string, HostFunction> = {}) {
    this.runtimeLimits = { ...DEFAULT_LIMITS, ...limits };
    for (const [name, fn] of Object.entries(functions)) this.register(name, fn);
  }

  /** Registers one host function after validating its name. */
  register(name: string, fn: HostFunction): void {
    if (!/^[A-Za-z_][A-Za-z0-9_]*$/.test(name)) throw new Error(`${JSON.stringify(name)} is not an identifier`);
    if (builtins.has(name)) throw new Error(`${name} is a built-in function`);
    this.hostFunctions.set(name, fn);
  }

  /** Returns the active resource limits. */
  limits(): Limits {
    return this.runtimeLimits;
  }

  /** Returns one registered host function. */
  hostFunction(name: string): HostFunction | undefined {
    return this.hostFunctions.get(name);
  }

  /** Registers one logical class function used by `Class::method(...)`. */
  registerClass(className: string, method: string, fn: HostFunction): void {
    if (!/^[A-Za-z_][A-Za-z0-9_]*$/.test(className) || !/^[A-Za-z_][A-Za-z0-9_]*$/.test(method)) {
      throw new Error('class function names must be identifiers');
    }
    this.classFunctions.set(`${className}::${method}`, fn);
  }

  /** Returns one logical class function. */
  classFunction(className: string, method: string): HostFunction | undefined {
    return this.classFunctions.get(`${className}::${method}`);
  }
}
